jwst/badpix_selfcal/badpix_selfcal_step.py

```diff
@@ -0,0 +1,210 @@
+
+        # apply the flags to the science data
+        input_sci = badpix_selfcal.apply_flags(input_sci, bad_indices)
+
+        # apply the flags to the background data to be passed to background sub step
+        if len(bkg_list) > 0:
+            for i, background_model in enumerate(bkg_list):
+                bkg_list[i] = badpix_selfcal.apply_flags(dm.open(background_model), bad_indices)
+
+        if self.save_flagged_bkg:
+            self.save_bkg(bkg_list)
+
+        input_sci.meta.cal_step.badpix_selfcal = 'COMPLETE'
+        return input_sci, bkg_list
+
+
+def _parse_inputs(input, selfcal_list, bkg_list):
+    """
+    Parse the input to the step. This is a helper function that is used in the
+    command line interface to the step.
+
+    Parameters
+    ----------
+    input: str
+        input science exposure or association
+
+    selfcal_list: list of ImageModels or filenames, default None
+        exposures to include as part of median background model used to find bad pixels,
+        but that are not flagged and returned as background exposures
+
+    bkg_list: list of ImageModels or filenames, default None
+        exposures to include as part of median background model used to find bad pixels,
+        and that are flagged and returned as background exposures
+
+    Returns
+    -------
+    input: JWST data model or association
+        input science data to be corrected
+
+    selfcal_list: list of ImageModels or filenames to use for selfcal
+    """
+    if selfcal_list is None:
+        selfcal_list = []
+    selfcal_list = [dm.open(selfcal) for selfcal in selfcal_list]
+    if bkg_list is None:
+        bkg_list = []
+    bkg_list = [dm.open(bkg) for bkg in bkg_list]
+    selfcal_list = selfcal_list + bkg_list
+
+    with dm.open(input) as input_data:
+
+        # find science and background exposures in association file
+        if isinstance(input_data, dm.ModelContainer):
+
+            sci_models, bkg_list_asn, selfcal_list_asn = split_container_by_asn_exptype(
+                input_data, exptypes=['science', 'background', 'selfcal'])
+
+            selfcal_list = selfcal_list + list(bkg_list_asn) + list(selfcal_list_asn)
+            bkg_list += list(bkg_list_asn)
+
+            if len(sci_models) > 1:
+                raise ValueError("Input data contains multiple science exposures. "
+                                 "This is not supported in calwebb_spec2 steps.")
+            input_sci = sci_models[0]
+
+        elif isinstance(input_data, dm.IFUImageModel) or isinstance(input_data, dm.ImageModel):
+
+            input_sci = input_data
+
+        else:
+            raise TypeError("Input data is not a ModelContainer, ImageModel, or IFUImageModel. "
+                            "Cannot continue.")
+
+    return input_sci, selfcal_list, bkg_list
+
+
+def split_container_by_asn_exptype(container: dm.ModelContainer, exptypes: list) -> list:
```

Comment:
This might be useful as a general utility function, perhaps over in `lib` somewhere.

Reply:
See conversation with @braingram above, probably marked as resolved now.  We discussed this possibility and decided there was no point in doing anything general because this will be replaced with ModelLibrary at some point in the future.  I have already asked for the ability to do this as part of what eventually gets shipped with the implementation of ModelLibrary

Reply:
See: https://github.com/spacetelescope/jwst/pull/8500#discussion_r1633685815

Reply:
You guys are way ahead of me!